### DISCUSSION TITLE
Request to Re-export IntoConflictValueClause and UndecoratedInsertRecord to `diesel::upsert`

### DISCUSSION THREAD

**Discussion by Abhicodes-crypto** [diesel-rs/diesel#3738] [Ideas]
While using `on_conflict_do_nothing()` function under a generic function. I get the error that ` IntoConflictValueClause and UndecoratedInsertRecord` trait bounds are not satisfied. These traits are declared pub, but query_builder::upsert is a private module and hence these traits can't be imported. 

It would be helpful if these traits are available for applications to import.

**Comment by Abhicodes-crypto** [comment]
Update : I am able to use this `UndecoratedInsertRecord` from diesel insert. I believe `IntoConflictValueClause` can also be made pub

**Comment by weiznich** [comment]
Types/Traits that are not publicly exported are private because we do not want to expose too much internals of diesel as stable API. This applies here as well. If you want to see a type as part of the public API I would expect such a proposal to include information about:

* Why this is necessary
* What alternatives there are (type aliases, other trait bounds, …)
* What implications it has in terms of what could or couldn't we change at a later point in time due to this change.

**Reply by LucaCappelletti94** [reply]
Hi,

While I cannot say what the OP might have needed this, I have encountered the same problem in my own use case while I was implement an `upsert` generic method in [`diesel-builders`](https://github.com/LucaCappelletti94/diesel-builders/).

Basically, I need to be able to express the constraints necessary to use [`on_conflict`](https://docs.diesel.rs/2.0.x/diesel/query_builder/struct.InsertStatement.html#method.on_conflict), but the `on_conflict` method is not part on an interface but just implemented for the following:

```rust
impl<T, U, Op, Ret> InsertStatement<T, U, Op, Ret>
where
    T: QuerySource,
    U: UndecoratedInsertRecord<T> + IntoConflictValueClause,
```

Where `U` is generally a `<X as Insertable<T>>::Values`. So, to be able to use `.on_conflict` in the following context:

```rust
diesel::insert_into(table)
    .values(values)
    .on_conflict(..)
```

I would need to specify the bound `<X as Insertable<T>>::Values: UndecoratedInsertRecord<T> + IntoConflictValueClause`, where `UndecoratedInsertRecord` is currently behind the `i-implement-a-third-party-backend-and-opt-into-breaking-changes` and `IntoConflictValueClause` is plainly private.

In my specific use case, `Values` is a variadic n-uple of `(diesel::dsl::Eq<Head, &'a Head::ColumnType>, ...)`. There was no problem defining the [generic insert query](https://github.com/LucaCappelletti94/diesel-builders/blob/5555a6c5e6bf43af31e35a7b1f838e51133d3689/diesel-builders/src/builder_bundle/completed_table_builder_bundle.rs#L216-L226) for such type, so I would have expected the same for the `upsert` variant.

Maybe we could move `on_conflict` to a `OnConflictDsl` trait of sorts?

It would not break any pre-existing APIs.

Best,
Luca

**Reply by weiznich** [reply]
Could you provide an error message showing what happens if you miss this bounds? Often there are hints in there that allow expression the necessary bound with the existing public API

**Reply by LucaCappelletti94** [reply]
Here it is:

```rust
error[E0599]: the method `on_conflict` exists for struct `InsertStatement<<M as HasTable>::Table, ...>`, but its trait bounds were not satisfied
   --> diesel-builders/src/ancestors.rs:291:14
    |
289 |           let results: Vec<<Self::Table as TableExt>::Model> = diesel::insert_into(table)
    |  ______________________________________________________________-
290 | |             .values(columns.eq_all(self.get_nested_columns()).flatten())
291 | |             .on_conflict(table.primary_key())
    | |             -^^^^^^^^^^^ method cannot be called on `InsertStatement<<M as HasTable>::Table, ...>` due to unsatisfied trait bounds
    | |_____________|
    |
    |
    = note: the following trait bounds were not satisfied:
            `<<<<<<M as HasTable>::Table as Table>::AllColumns as tuplities::prelude::NestTuple>::Nested as TupleEqAll>::EqAll as tuplities::prelude::FlattenNestedTuple>::Flattened as Insertable<<M as HasTable>::Table>>::Values: diesel::query_builder::upsert::into_conflict_clause::IntoConflictValueClause`
            `<<<<<<M as HasTable>::Table as Table>::AllColumns as tuplities::prelude::NestTuple>::Nested as TupleEqAll>::EqAll as tuplities::prelude::FlattenNestedTuple>::Flattened as Insertable<<M as HasTable>::Table>>::Values: UndecoratedInsertRecord<<M as HasTable>::Table>`
    = note: the full name for the type has been written to '/home/lucac/github/diesel-builders/target/debug/deps/diesel_builders-beee45d7e7cc917d.long-type-11839854388964568874.txt'
    = note: consider using `--verbose` to print the full type name to the console
```

With PR #4914, this can now be expressed with:

```rust
InsertStatement<
        Self::Table,
        <<<<<M::Table as Table>::AllColumns as NestTuple>::Nested as TupleEqAll>::EqAll as FlattenNestedTuple>::Flattened as Insertable<Self::Table>>::Values,
    >: OnConflictDsl<<M::Table as Table>::PrimaryKey>
```
